L82-core-event-engine.md

```diff
@@ -0,0 +1,564 @@
+
+Unlike the current polling model where threads are borrowed from the
+application, EventEngines will bring their own polling and callback execution
+threads. This public API will make it easier to integrate gRPC into external
+event loops, it will support siloing events between gRPC channels and servers,
+and it will provide another way to support the C++ Callback API.
+
+## Background
+
+gRPC Core and its wrapped languages rely internally on an event management
+framework called iomgr. A small collection of homegrown, platform-specific
+implementations exist, allowing I/O functionality to be customized to the
+various platforms that gRPC supports (IOCP is used on Windows, epoll on Linux,
+poll on Mac, etc).
+
+Over the past few years, there have been multiple projects aimed at improving or
+replacing iomgr. The EventEngine effort takes a slightly different approach to
+things, but otherwise shares the same rationale as the previous effort.
+
+### Related Proposals:
+
+* This work supersedes [L69: EventManger API](https://github.com/grpc/proposal/pull/182)
+* Supports [L67: C++ callback-based asynchronous API](https://github.com/grpc/proposal/pull/180) as a more performant solution than provided by the [CallbackAlternativeCQ](https://github.com/grpc/grpc/pull/25169).
+
+## Proposal
+
+### Concepts
+
+An EventEngine is roughly composed of:
+
+* A task runner which asynchronously executes callables with optional
+  scheduling.
+* An asynchronous DNS resolver.
+* Network listener logic that accepts incoming connection requests, and
+  performs network-server-related I/O operations.
+* Client logic, support for making outgoing network connections.
+
+### Common Data Types
+
+```cpp
+namespace grpc_event_engine {
+namespace experimental {
+
+/// A set of parameters used to configure an endpoint, either when initiating a
+/// new connection on the client side or when listening for incoming connections
+/// on the server side. An EndpointConfig contains a set of zero or more
+/// Settings. Each setting has a unique name, which can be used to fetch that
+/// Setting via the Get() method. Each Setting has a value, which can be an
+/// integer, string, or void pointer. Each EE impl should define the set of
+/// Settings that it supports being passed into it, along with the corresponding
+/// type.
+class EndpointConfig {
+ public:
+  virtual ~EndpointConfig() = default;
+  using Setting = absl::variant<absl::monostate, int, absl::string_view, void*>;
+  /// Returns an EndpointConfig Setting. If there is no Setting associated with
+  /// \a key in the EndpointConfig, an \a absl::monostate type will be
+  /// returned. Caller does not take ownership of resulting value.
+  virtual Setting Get(absl::string_view key) const = 0;
+};
+```
+
+The `SliceAllocator` provides memory buffers for EventEngine read/write
+operations, allowing EventEngines to seamlessly participate in the
+C++ `ResourceQuota` system.
+
+```cpp
+class SliceAllocator {
```

Comment:
Could we turn this into an interface perhaps?

I could imagine a future pivot where we implement SliceAllocator directly in resource user - but also a nearer future where we increase our testability by substituting in a fake SliceAllocator to verify behavior of EventEngine implementations.

Reply:
@markdroth, please correct me if wrong, but I believe the main justification against an interface was that we wanted to avoid dynamic allocation in the Listener and Endpoint APIs where possible. Do you have any other objections here?

Reply:
Discussed this externally on another PR, seems like there are no objections. I'll make this a pure virtual interface.

Reply:
Updated, PTAL